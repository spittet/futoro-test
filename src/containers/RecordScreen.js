/**
 * This is the screen where the user records a capsule
 *
 * BIG WARNING! The timeout is likely to fail if you're using Chrome Debug.
 * Check https://github.com/facebook/react-native/issues/4470 for more details.
 * Instead you should run it on your mobile and everythign should be fine.
 *
 * @flow
 */

import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { View } from 'react-native';
import Camera from 'react-native-camera';
import LCTouchableImage from '../components/LCTouchableImage'
import styles from './styles';

import { config } from '../config';
import { NavigationActions } from 'react-navigation';
import { 
  cancelNewCapsule,
  recordNewCapsule, 
} from '../actions';
import type Capsule from '../reducers';

class RecordScreen extends React.Component {

  camera: any;
  
  // Not using Redux global state because we don't need this data elsewhere.
  state: {
    camera: {
      captureAudio: boolean,
      captureMode: number,
      aspect: number,
      captureTarget: number,
      type: number,
      orientation: number,
      flashMode: number,
      captureQuality: number
    },
    isRecording: boolean,
    recordStoppedByUser: boolean,
    recordingTimeoutID: ?number
  };

  static navigationOptions = {
    title: 'Record Screen',
  }

  constructor(props: Object) {
    super(props);
    
    this.camera = null;

    this.state = {
      camera: {
        captureAudio: true,
        captureMode: Camera.constants.CaptureMode.video,
        aspect: Camera.constants.Aspect.fill,
        captureTarget: Camera.constants.CaptureTarget.temp,
        type: Camera.constants.Type.back,
        orientation: Camera.constants.Orientation.auto,
        flashMode: Camera.constants.FlashMode.auto,
        captureQuality: Camera.constants.CaptureQuality.medium
      },
      isRecording: false,
      recordStoppedByUser: false,
      recordingTimeoutID: null
    };
  }


  componentWillUnmount() {
    // It seems that the camera automatically stop capturing when Unmounting
    //this.cancelRecording();
  }

  // Sets a timeout after which the recording is automatically stop
  _setRecordingTimeout = () => {
    let recordingTimeoutID = setTimeout(() => {
      this._stopRecording();
    }, config.MAX_RECORDING_DURATION);

    this.setState({
      recordingTimeoutID: recordingTimeoutID
    });
  }

  // Clears the recording timeout to prevent leaks
  _clearRecordingTimeout = () => {
    this.state.recordingTimeoutID &&
      clearTimeout(this.state.recordingTimeoutID);
  }

  _startRecording = () => {
    if (this.camera && !this.state.isRecording) {
      let captureOptions = {
        mode: this.state.camera.captureMode,
        audio: this.state.camera.captureAudio
      }

      this.camera.capture(captureOptions)
        .then(async (data) => { // Record new video to global state

          this._clearRecordingTimeout();

          let capsule: Capsule = {
            uri: data.path
          };
          
          // If the users presses the back button we don't record video
          if (this.state.recordStoppedByUser) {
            await this.props.dispatch(recordNewCapsule(capsule));

            // Automatically redirects to the Preview page
            this.props.dispatch(
              NavigationActions.navigate({ routeName: 'Preview' })
            ); 
          } else {  // This happens if user presses the back button
            this.props.dispatch(cancelNewCapsule(capsule));
          }
        });

      this._setRecordingTimeout();
      
      this.setState({
        isRecording: true,
        recordStoppedByUser: false
      });    
    }
  }

  _stopRecording = () => {
    if (this.camera && this.state.isRecording) {
      this.setState({ 
        isRecording: false,
        recordStoppedByUser: true   // Video 
      }, () => {
        this.camera.stopCapture();
      });
      
    }
  }

  render() {
    return (
      <View style={styles.container}>
        <Camera
          ref={(cam) => {
            this.camera = cam;
          }}
          style={styles.preview}
          aspect={this.state.camera.aspect}
          captureTarget={this.state.camera.captureTarget}
          type={this.state.camera.type}
          flashMode={this.state.camera.flashMode}
          captureAudio={this.state.camera.captureAudio}
          captureQuality={this.state.camera.captureQuality}
          onFocusChanged={() => {}}
          onZoomChanged={() => {}}
          defaultTouchToFocus
          mirrorImage={false}
        />
        <View style={[styles.overlay, styles.bottomOverlay]}>
          <View style={styles.buttonsSpace} />
          {
            !this.state.isRecording
            &&
            <LCTouchableImage
              id="startRecordingButton"
              buttonStyle={styles.captureButton}
              buttonAction={this._startRecording}
              imageSrc={require('../assets/ic_videocam_36pt.png')}
            />
            ||
            <LCTouchableImage 
              id="stopRecordingButton"
              buttonStyle={styles.captureButton}
              buttonAction={this._stopRecording}
              imageSrc={require('../assets/ic_stop_36pt.png')}
            />
          }
        </View>
      </View>
    );
  }
}

RecordScreen.propTypes = {
  dispatch: PropTypes.func.isRequired,
  newCapsule: PropTypes.object
}

const mapStateToProps = (state) => {
  return {
    newCapsule: state.newCapsule
  }
}


export default connect(mapStateToProps)(RecordScreen);
