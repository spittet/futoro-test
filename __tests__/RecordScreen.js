/* eslint-env mocha */
/* eslint-disable */

import 'react-native';
import React from 'react';
//import ReactTestUtils from 'react-dom/test-utils';
import { shallow } from 'enzyme';
import RecordScreen from '../src/containers/RecordScreen.js';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';

jest.mock('react-native-fs', () => 'RNFS');

describe('Testing the rendering of the recording Screen', () => {
  it('renders default component as expected', () => {
    const wrapper = shallow(
      <RecordScreen />
    );
    expect(wrapper).toMatchSnapshot();
  });
  it('renders the stop button when it\'s recording', () => {
    const wrapper = shallow(
      <RecordScreen />
    );
    wrapper.setState({isRecording: true});
    expect(wrapper.find({id: 'stopRecordingButton'})).toHaveLength(1);
  });
});