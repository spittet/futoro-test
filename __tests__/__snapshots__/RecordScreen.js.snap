// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing the rendering of the recording Screen renders default component as expected 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Camera
    defaultTouchToFocus={true}
    mirrorImage={false}
    onFocusChanged={[Function]}
    onZoomChanged={[Function]}
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "flex-end",
      }
    }
  />
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "left": 0,
          "padding": 16,
          "position": "absolute",
          "right": 0,
        },
        Object {
          "alignItems": "center",
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "top": 0,
        },
      ]
    }
  >
    <LCTouchableImage
      buttonAction={[Function]}
      buttonStyle={
        Object {
          "padding": 5,
        }
      }
      id="switchTypeButton"
      imageSrc={1}
    />
    <LCTouchableImage
      buttonAction={[Function]}
      buttonStyle={
        Object {
          "padding": 5,
        }
      }
      id="switchFlashButton"
      imageSrc={1}
    />
  </View>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "left": 0,
          "padding": 16,
          "position": "absolute",
          "right": 0,
        },
        Object {
          "alignItems": "center",
          "backgroundColor": "rgba(0,0,0,0.4)",
          "bottom": 0,
          "flexDirection": "row",
          "justifyContent": "center",
        },
      ]
    }
  >
    <View
      style={
        Object {
          "width": 10,
        }
      }
    />
    <LCTouchableImage
      buttonAction={[Function]}
      buttonStyle={
        Object {
          "backgroundColor": "white",
          "borderRadius": 40,
          "padding": 15,
        }
      }
      id="startRecordingButton"
      imageSrc={1}
    />
  </View>
</View>
`;
